src/devices/JkBmsV13.cpp

```diff
@@ -0,0 +1,292 @@
+    //mqttPublish("bms/"+String(BT_DEVICES_COUNT+u8_mDevNrJkV13)+"/totalVoltage", getBmsTotalVoltage(BT_DEVICES_COUNT+u8_mDevNrJkV13));
+    //mqttPublish("bms/"+String(BT_DEVICES_COUNT+u8_mDevNrJk)+"/totalCurrent", getBmsTotalCurrent(BT_DEVICES_COUNT+u8_mDevNrJk));
+  }
+  else
+  {
+    ESP_LOGD(TAG_V13,"bmsData checksum wrong; Serial(%i)\n",u8_mDevNrJkV13);
+    bo_lRet=false;
+  }
+
+
+  if(bo_lRet==false) return bo_lRet;
+  
+  setBmsLastDataMillis(BT_DEVICES_COUNT+u8_mDevNrJkV13,millis());
+
+  return bo_lRet;  
+}
+
+
+void JkBmsV13_sendMessage(uint8_t *sendMsg , uint32_t size)
+{
+  if(u8_mTxEnRS485pinJkV13>0) digitalWrite(u8_mTxEnRS485pinJkV13, HIGH); 
+  usleep(50);
+  mPortJkV13->write(sendMsg, size);
+  mPortJkV13->flush();  
+  //usleep(1000);
+  if(u8_mTxEnRS485pinJkV13>0) digitalWrite(u8_mTxEnRS485pinJkV13, LOW); 
+}
+
+
+bool JkBmsV13_recvAnswer(uint8_t *p_lRecvBytes)
+{
+  uint8_t SMrecvState, u8_lRecvByte;
+  uint16_t u16_lRecvDataLen;
+  uint32_t u32_lStartTime=millis();
+  SMrecvState=SEARCH_START_BYTE1;
+  u16_mLastRecvBytesCntJkV13=0;
+  uint8_t crc=0;
+  char strTmp[1024];
+  char *sptr = strTmp;
+
+  for(;;)
+  {
+    //Timeout
+    if(millis()-u32_lStartTime > 500) 
+    {
+      ESP_LOGD(TAG_V13,"Timeout: Serial=%i, u8_lRecvDataLen=%i, u8_lRecvBytesCnt=%i\n",u8_mDevNrJkV13, u16_lRecvDataLen, u16_mLastRecvBytesCntJkV13);
+      //for(uint16_t x=0;x<u16_lRecvDataLen;x++)
+      //{
+      //  ESP_LOGD(TAG_V13,"%02x ",p_lRecvBytes[x]);
+      //}
+      if(u16_lRecvDataLen != 0){
+        ESP_LOGD(TAG_V13,"%s",strTmp);
+      }
+      return false;
+    }
+
+    //Überprüfen ob Zeichen verfügbar
+    if (mPortJkV13->available() > 0)
+    {
+      u16_lRecvDataLen++;
+      u8_lRecvByte = mPortJkV13->read();
+
+      //ESP_LOGD(TAG_V13,"%02x ",u8_lRecvByte);
+      sprintf(sptr,"%02x ",u8_lRecvByte);
+      sptr += 3;
+
+      if(u16_mLastRecvBytesCntJkV13<JKBMSV13_MAX_ANSWER_LEN-1){crc += u8_lRecvByte;}
+
+      switch (SMrecvState)  {
+        case SEARCH_START_BYTE1:
+          if (u8_lRecvByte == 0xEB){SMrecvState=SEARCH_START_BYTE2;}
+          break;
+        case SEARCH_START_BYTE2:
+          if (u8_lRecvByte == 0x90){SMrecvState=SLAVE_ADDR;}
+          break;
+        case SLAVE_ADDR:
+          if (u8_lRecvByte == JK_V13_SLAVE_ADDR){SMrecvState=CMD_CODE;}
+          break;   
+        case CMD_CODE:
+          if (u8_lRecvByte == 0xFF){SMrecvState=RECV_DATA;}
+          break;                      
+        case RECV_DATA:
+          p_lRecvBytes[u16_mLastRecvBytesCntJkV13]=u8_lRecvByte;
+          u16_mLastRecvBytesCntJkV13++;
+          break;      
+        default:
+          SMrecvState=SEARCH_START_BYTE1;
+          crc = 0;
+          break;
+        }
+
+    }
+
+    if(u16_mLastRecvBytesCntJkV13==JKBMSV13_MAX_ANSWER_LEN) break; //Recv Pakage complete
+  }
+
+  //Überprüfe Cheksum
+  //debugPrintf("crc=%i %i\n", crcB3, crcB4);
+  if(p_lRecvBytes[u16_mLastRecvBytesCntJkV13-1]!=crc) return false; 
```

Comment:
Eine Berechung des CRC wäre meiner Meinung nach nicht verkehrt.
Aus Erfahrung (beim JBD BMS) weiß ich, dass sporadisch einmal ein Packet dabei ist, in dem falsche Werte enthalten sind. Das fällt dann durch die CRC-Prüfung auf.
Kann man aber auch später noch machen.

Reply:
Ja CRC ist schon wichtig ... ich sehs grad ned was nicht passt, die wird berechnet und die Antwort verworfen wenn was nicht richtig ist.